### PR TITLE
Fix internal error of a single case active pattern

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

Fixes # https://github.com/dotnet/fsharp/issues/16856

### Before
`internal error: Unexpected generalized type variables when compiling active pattern.`

### After

`(Error 3188, Line 3, Col 6, Line 3, Col 7, "Type inference caused an inference type variable to escape its scope. Consider adding type annotations to make your code less generic.")`

## Checklist

- [x] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by edgarfgp** [comment]
Based on the CI result and the test added. 
- I think we could remove the check for  `not (isNil origInputValTypars)` that checks for 
`Unexpected generalized type variables when compiling an active pattern`. 
-  Removing this check still raises a compiler error `Type inference caused an inference type variable to escape its scope. Consider adding type annotations to make your code less generic` which will guide the user to add the needed type annotations to make the code compile. See Tests in this PR

Any thoughts? @T-Gro @vzarytovskii ?

File: src/Compiler/Checking/PatternMatchCompilation.fs
Comment (psfinaki) [inline src/Compiler/Checking/PatternMatchCompilation.fs:1298]: I tried to blame this line but looks like it's been there forever :shrug:

**Review by psfinaki — APPROVED** [review]
Looks like a valid change for me. The new behavior is reasonable I think. Great testing as usual :)

**Comment by edgarfgp** [comment]
> Looks like a valid change for me. The new behavior is reasonable I think. Great testing as usual :)

Yeah ,this check stops the rest of the analysis

**Comment by edgarfgp** [comment]
Can I get a second review here please ?

**Comment by psfinaki** [comment]
Most team members are still on Easter vacations :) 
Maybe @T-Gro

**Review by T-Gro — CHANGES_REQUESTED** [review]
Hi Edgar.

The fix probably needs better understanding of what is happening in the error situations and why.

The fact that changing ` _ -> _ ` to `int -> _` fixes is is suspicios.
Also, the error message after the fix is not really good either, it would for sure better to better understand the type generalization scenario and apply a proper error message based on it.

I am marking this for now to prevent accidental merging.

**Comment by edgarfgp** [comment]
> The fact that changing _ -> _ to int -> _

@T-Gro  This happens with or without the internal error.
 I agree that there is something happening on the `PatternMatchingCompilation` analysis. 

The aim of this PR is allow the analysis to continue and let the type checking guide the user.

Im not keen of going deep on the `PatternMatchingCompilation` tried in the past and the analysis was fragile at some points.

Another option is to keep the check and remove the `InternalError` and use a proper error message and number.

```fsharp
if not (isNil origInputValTypars) then errorR(MyNewError("Unexpected generalized type variables when compiling an active pattern", m))
```

**Review by KevinRansom — CHANGES_REQUESTED** [review]
@edgarfgp --
That InternalError has been in the codebase forever, we need to dig in and understand what the intent of it is.  That is going to take some time and effort.  Removing the check, relying on a subsequent analysis error may be the correct approach, but without digging in we won't be confident.  We mostly like to be confident about changes we merge.

**Comment by edgarfgp** [comment]
> @edgarfgp -- That InternalError has been in the codebase forever, we need to dig in and understand what the intent of it is. That is going to take some time and effort. Removing the check, relying on a subsequent analysis error may be the correct approach, but without digging in we won't be confident. We mostly like to be confident about changes we merge.

Another option is to keep the check and remove the InternalError and use a proper error message and number

**Comment by KevinRansom** [comment]
@edgarfgp is that better? or the same with a less urgent error?  My preference is that we explain why the check is necessary and then fix the message based on that understanding or eliminate it, otherwise a change is just churn.

**Comment by edgarfgp** [comment]
> @edgarfgp is that better? or the same with a less urgent error? My preference is that we explain why the check is necessary and then fix the message based on that understanding or eliminate it, otherwise a change is just churn.

Fair. Just that cannot  commit to that investigation as of now.  Will close this PR for now and reopening when I have more time.

BTW Thanks for the review

**Comment by KevinRansom** [comment]
@edgarfgp thanks mate, I appreciate it.

**Comment by psfinaki** [comment]
Just to throw my perspective in here - in this cases we have only few things to do:
- git blame
- scrutinize 
- <s>hope and pray</s> rely on our tests

I've blamed the code as far as I could, but yeah the line was checked in 10 years ago together with most of the logic here by latkin. 
So unless somewhat has some "a-ha" insight when looking at this, we can't do much.

With the current release cadence and previews and all, we can afford ourselves not to keep things there **just** because of paranoia. The fact nothing fails is good news.

TLDR: let's scrutinize that and - if we don't come with anything - let's merge this.

**Comment by edgarfgp** [comment]
IMO `internal compiler errors` should not show to the end users it is just bad look

**Comment by vzarytovskii** [comment]
> IMO `internal compiler errors` should not be shown to the end users it is just bad look

It's orthogonal. If it's shown, it means something exceptional happened, and we need to figure out what and why. Just removing a check, while it works, doesn't help understanding what's wrong.

**Comment by edgarfgp** [comment]
> > IMO `internal compiler errors` should not be shown to the end users it is just bad look
> 
> It's orthogonal. If it's shown, it means something exceptional happened, and we need to figure out what and why. Just removing a check, while it works, doesn't help understanding what's wrong.

@vzarytovskii I’m not arguing that. Is more about these stop the rest of the code analysis. Hopefully someone from the team can look into this at some point

**Comment by vzarytovskii** [comment]
> > > IMO `internal compiler errors` should not be shown to the end users it is just bad look
> 
> > 
> 
> > It's orthogonal. If it's shown, it means something exceptional happened, and we need to figure out what and why. Just removing a check, while it works, doesn't help understanding what's wrong.
> 
> 
> 
> @vzarytovskii I’m not arguing that. Is more about these stop the rest of the code analysis. Hopefully someone from the team can look into this at some point 

I can tomorrow when I'm back from the vacation, let me set the reminder

**Comment by vzarytovskii** [comment]
> Did the reminder work 😁?

It actually did, but then I fully switched to the VS crashing issue and some non-working Roslyn integration, so shelved this one.

**Comment by edgarfgp** [comment]
> > Did the reminder work 😁?
> 
> It actually did, but then I fully switched to the VS crashing issue and some non-working Roslyn integration, so shelved this one.

No worries. Will try and have a look at some point during the week. I really want the find a solution for this `internal error`.

**Pull request closed without merging**
